### PR TITLE
[spiral/core] Add the ability to use `BackedEnum` as a scope name

### PR DESCRIPTION
**Pull request by msmakouz**
| Q             | A
| ------------- | ---
| Bugfix?       | ❌
| Breaks BC?    | ❌ 
| New feature?  | ✔️

Added the ability to use `BackedEnum` as a scope name. Introduced the `Spiral\Framework\ScopeName` enum with default scopes.

```php
enum ScopeName: string
{
    case Http = 'http';
    case HttpRequest = 'http.request';
    case Queue = 'queue';
    case QueueTask = 'queue.task';
    case Temporal = 'temporal';
    case TemporalActivity = 'temporal.activity';
    case Grpc = 'grpc';
    case GrpcRequest = 'grpc.request';
    case Centrifugo = 'centrifugo';
    case CentrifugoRequest = 'centrifugo.request';
    case Tcp = 'tcp';
    case TcpPacket = 'tcp.packet';
    case Console = 'console';
    case ConsoleCommand = 'console.command';
}
```

### CODE REVIEW

File: src/Framework/Framework/ScopeName.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Spiral\Framework;
+
+enum ScopeName: string
+{
+    case HttpRequest = 'http.request';
```
Comment (butschster) [inline src/Framework/Framework/ScopeName.php:9]: There are also `http`, `queue`, `temporal`, `grpc`, `tcp`, `console` cases

**Review by butschster — CHANGES_REQUESTED** [review]

File: src/Core/src/Attribute/Scope.php [hunk 2]
```diff
@@ -12,8 +12,10 @@
 #[\Attribute(\Attribute::TARGET_CLASS)]
 final class Scope implements Plugin
 {
-    public function __construct(
-        public string $name,
-    ) {
+    public string $name;
```
Comment (roxblnfk) [inline src/Core/src/Attribute/Scope.php:15]:
```suggestion
    public readonly string $name;
```

File: src/Core/src/Container.php [hunk 3]
```diff
@@ -153,13 +153,15 @@ public function has(string $id): bool
...
-     * @param null|string $scope Scope name.
+     * @param null|\BackedEnum|string $scope Scope name.
      *        If {@see null}, binder for the current working scope will be returned.
      *        If {@see string}, the default binder for the given scope will be returned. Default bindings won't affect
      *        already created Container instances except the case with the root one.
      */
-    public function getBinder(?string $scope = null): BinderInterface
+    public function getBinder(string|\BackedEnum|null $scope = null): BinderInterface
     {
+        $scope = $scope instanceof \BackedEnum ? $scope->value : $scope;
```
Comment (roxblnfk) [inline src/Core/src/Container.php:163]: Check like `is_object` or `is_enum` might be faster than `instanceof`

File: src/Core/src/Attribute/Scope.php [hunk 4]
```diff
@@ -12,8 +12,10 @@
...
+
+    public function __construct(string|\BackedEnum $name)
+    {
+        $this->name = $name instanceof \BackedEnum ? $name->value : $name;
```
[the remark this answers is not in the document]
Comment (msmakouz) [reply src/Core/src/Attribute/Scope.php]: Added.

File: src/Core/src/Attribute/Scope.php (hunk 2 above)
Comment (msmakouz) [reply src/Core/src/Attribute/Scope.php]: Added.

File: src/Core/src/Container.php (hunk 3 above)
Comment (msmakouz) [reply src/Core/src/Container.php]: Changed to is_object

File: src/Core/src/Container.php [hunk 5]
```diff
@@ -379,7 +381,7 @@ private function closeScope(): void
     private function runIsolatedScope(Scope $config, callable $closure): mixed
     {
         // Open scope
-        $container = new self($this->config, $config->name);
+        $container = new self($this->config, \is_object($config->name) ? (string) $config->name->value : $config->name);
```
Comment (roxblnfk) [inline src/Core/src/Container.php:384]: might be moved in the constructor

Comment (msmakouz) [reply src/Core/src/Container.php]: Moved

**Pull request merged**